<a href="https://colab.research.google.com/github/cristobalperezp/CC6205-NLP/blob/main/Tarea_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:** Marcelo Rojas y Cristóbal Pérez.

**Fecha límite de entrega 📆:** Martes 4 de Abril.

**Tiempo estimado de dedicación:** 




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 100 palabras** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones según lo visto en clases. Debe citar la slide donde cree que está la respuesta.**




` ` 
Respuesta aquí
` ` 

**Pregunta 2 (0.4 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.


` ` 
Respuesta aquí
` ` 

 **Pregunta 3 (0.4 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 
Respuesta aquí
` ` 

**Pregunta 4 (0.4 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.

` ` 
Respuesta aquí
` ` 

**Pregunta 5 (0.6 puntos): Acorde al paper [A Vector Space Model for automatic indexing](https://dl.acm.org/doi/pdf/10.1145/361219.361220) un documento, $D_i$, puede ser definido formalmente como una tupla de términos, $(d_{i1}, d_{i2}, ..., d_{in})$, donde $d_{ij}$ representa el paso del j-esimo término. En clase vieron algunas formas medir los pesos de estos términos. Mencione cuales fueron y sus ventajas y desventajas.** 

` ` 
Respuesta aquí
` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Tokens*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [ ]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [ ]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [ ]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

In [ ]:
try:
    from google.colab import drive

    drive.mount("/content/drive")
    path = 'path/to/marcianeke.txt'
except: 
    print('Ignorando conexión drive-colab')

Mounted at /content/drive


` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver qué tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [ ]:
! git clone 'https://github.com/cristobalperezp/CC6205-NLP.git'
text = codecs.open('/content/CC6205-NLP/marcianeke.txt', 'r', 'UTF-8').read()
print(text)

fatal: destination path 'CC6205-NLP' already exists and is not an empty directory.
Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [ ]:
#def get_tokens(text):
  ### Inicio del código ###
  
#  import string
#  text = re.sub(r'[^\w\s]',' ', text)
#  text = text.split()
#  return text
  
  ### Fin del código ###


def get_tokens(text):
  ### Inicio del código ###
  spaces = ['',' ','\n','\r']
  punctuation = [',','.','-','(',')',':',';',"'"]
  tokens = []
  aux = ''
  for i in text:
    if i in spaces:
      if aux!='':
        tokens.append(aux)
        aux = ''
    elif i in punctuation:
      if aux!='':
        tokens.append(aux)
        aux = ''
      tokens.append(i)
    else:
      aux = aux + i
  if aux!='':
        tokens.append(aux)
  return tokens
  ### Fin del código ###

In [ ]:
tokens = get_tokens(text)
tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

#Supuestos aqui


In [ ]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

['Brr',
 'Marcianeke',
 'Vamo',
 "'",
 'a',
 'estar',
 'con',
 'Pailita',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 '(',
 'brr',
 ')',
 'Tussi',
 ',',
 'keta',
 'quiere',
 "'",
 'mezclar',
 'Dimelo',
 'má',
 'Ando',
 'en',
 'busca',
 'de',
 'una',
 'criminal',
 '(',
 'ah',
 ',',
 'ah',
 ')',
 'Esa',
 'que',
 'el',
 'gatillo',
 'le',
 'gusta',
 'jalar',
 '(',
 'rata',
 '-',
 'ta',
 ')',
 'Que',
 'le',
 'guste',
 'flotar',
 'y',
 'fumar',
 'Tussi',
 ',',
 'keta',
 'pura',
 'quiere',
 "'",
 'mezclar',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Di',
 '-',
 'dimelo',
 'má',
 'Esperame',
 'que',
 'ahora',
 'entro',
 'yo',
 'Y',
 'lo',
 'que',

In [ ]:
# Vemos en qué tokens difieren el tokenizador programado con el de nltk
diff = [x for x in nltk_tokens if x not in tokens]
diff = set(diff)
diff

set()

#En qué difieren:

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [ ]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los tokens de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [ ]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  vocab = []
  for doc in dataset:
    vocab += get_tokens(doc)
  vocab = list(set(vocab))
  vocab.sort()
  return vocab
  ### Aquí termina tu código ###

***Test:***

In [ ]:
vocab = get_vocab(dataset)
vocab

['I',
 'Spanish',
 'favorite',
 'human',
 'is',
 'language',
 'languages',
 'like',
 'my',
 'programming']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí: 

    Como el vocabulario está en inglés, se utilizará la lista de stopwords de 
    nltk para las reglas de stopwords, y por el lado de stemming se usarán las 
    reglas del algoritmo de Porter.

In [ ]:
def pre_processing(vocab):
  ### Aquí inicia tu código ###
  nltk_stopwords = ["i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him",
                    "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves",
                    "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be","been", "being","have",
                    "has", "had", "having", "do", "does", "did", "doing", "a", "an", "the", "and", "but", "if", "or","because", "as", "until", "while",
                    "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below",
                    "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", 
                    "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", 
                    "own", "same", "so", "than", "too", "very", "s", "t", "can", "will", "just", "don", "should", "now",'I','Me','My','Myself','We',
                    'Our','Ours','Ourselves','You','Your','Yours','Yourself','Yourselves','He','Him','His','Himself','She','Her','Hers','Herself','It',
                    'Its','Itself','They','Them','Their', 'Theirs','Themselves','What','Which','Who','Whom','This','That','These','Those', 'Am','Is','Are',
                    'Was','Were','Be','Been','Being','Have','Has','Had','Having','Do','Does','Did','Doing','A','An','The','And','But','If','Or','Because',
                    'As','Until','While','Of','At', 'By','For','With', 'About','Against','Between','Into','Through','During','Before','After','Above',
                    'Below','To','From','Up','Down','In','Out','On','Off','Over','Under','Again','Further','Then','Once','Here','There','When','Where',
                    'Why','How','All','Any','Both','Each','Few','More','Most','Other','Some','Such','No','Nor','Not','Only','Own','Same','So','Than',
                    'Too','Very','S','T','Can','Will','Just','Don','Should','Now']

  porter_rules = {'sses': 'ss', 'ies':'i', 's':''}
  for id, i in enumerate(vocab):
    if i in nltk_stopwords:
      vocab.remove(i)
    else:
      for j in list(porter_rules.keys()):
        if i.endswith(j):
          replace_word = i.replace(j, porter_rules[j])
          vocab[id] = replace_word
          break
  vocab = list(set(vocab))
  return vocab
  ### Aquí termina tu código ###
    

In [ ]:
vocab = pre_processing(vocab)
vocab

['programming', 'like', 'human', 'language', 'favorite', 'Spanish']


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [ ]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [ ]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  voc = get_vocab(dataset)
  bow = []
  for doc in dataset:
    dict_aux = dict.fromkeys(voc,0)
    words_doc = doc.split()
    for word in words_doc:
      dict_aux[word] = words_doc.count(word)
    bow.append(dict_aux)
  index = ['d'+str(i) for i in range(len(dataset))]
  bow = pd.DataFrame(bow,index=index)
  return bow
  ### Aquí termina tu código ###

In [ ]:
dataset = ['El perro ladra', 'El perro come']

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

El  a  come  comida  despierta  después  duerme  empieza  gato  la  ladra  \
d0   1  0     1       1          0        1       1        0     0   1      0   
d1   1  1     0       0          1        1       0        1     0   0      0   
d2   1  0     1       1          0        1       0        0     0   1      1   
d3   1  0     1       1          0        1       1        0     1   1      0   
d4   1  1     0       0          1        1       0        1     1   0      0   
d5   1  0     1       1          0        1       0        0     1   1      0   

    ladrar  maulla  maullar  perro  se  y  
d0       0       0        0      1   2  1  
d1       1       0        0      1   1  1  
d2       0       0        0      1   1  1  
d3       0       0        0      0   2  1  
d4       0       0        1      0   1  1  
d5       0       1        0      0   1  1

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
  ### Aquí inicia tu código ###
  n_docs = dataset_bow.shape[0]
  columnas = dataset_bow.columns
  datos = dataset_bow.to_numpy()
  tf = []
  for i in range(n_docs):
    tf.append(
        datos[i]/datos[i].max()
        )
  index = ['d'+str(i) for i in range(len(dataset_bow))]
  tf = pd.DataFrame(tf,columns=columnas,index=index)
  return tf
  ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

El    a  come  comida  despierta  después  duerme  empieza  gato   la  \
d0  0.5  0.0   0.5     0.5        0.0      0.5     0.5      0.0   0.0  0.5   
d1  1.0  1.0   0.0     0.0        1.0      1.0     0.0      1.0   0.0  0.0   
d2  1.0  0.0   1.0     1.0        0.0      1.0     0.0      0.0   0.0  1.0   
d3  0.5  0.0   0.5     0.5        0.0      0.5     0.5      0.0   0.5  0.5   
d4  1.0  1.0   0.0     0.0        1.0      1.0     0.0      1.0   1.0  0.0   
d5  1.0  0.0   1.0     1.0        0.0      1.0     0.0      0.0   1.0  1.0   

    ladra  ladrar  maulla  maullar  perro   se    y  
d0    0.0     0.0     0.0      0.0    0.5  1.0  0.5  
d1    0.0     1.0     0.0      0.0    1.0  1.0  1.0  
d2    1.0     0.0     0.0      0.0    1.0  1.0  1.0  
d3    0.0     0.0     0.0      0.0    0.0  1.0  0.5  
d4    0.0     0.0     0.0      1.0    0.0  1.0  1.0  
d5    0.0     0.0     1.0      0.0    0.0  1.0  1.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [ ]:
def calc_idf(dataset_bow):
  ### Aquí inicia tu código ###
  n_docs = dataset_bow.shape[0]
  n_words = dataset_bow.shape[1]
  df = np.zeros(n_words)
  for i in range(n_words):
    df[i] = np.sum((dataset_bow.iloc[:,i].values != 0)*1)
    #idf_i = np.log((n_docs+1)/(df+1)) +1

  idf_i = np.log10((n_docs)/(df))

  words = dataset_bow.columns
  idf_dict = dict(zip(words, idf_i))
  return idf_dict
  ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
  ### Aquí inicia tu código ###
  columnas = tf.columns
  tfidf = tf*idf.values()
  index = ['d'+str(i) for i in range(len(tf))]
  df = pd.DataFrame(tfidf, columns=columnas,index=index)
  return df

  ### Aquí termina tu código ### 

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El         a      come    comida  despierta  después    duerme   empieza  \
d0  0.0  0.000000  0.088046  0.088046   0.000000      0.0  0.238561  0.000000   
d1  0.0  0.477121  0.000000  0.000000   0.477121      0.0  0.000000  0.477121   
d2  0.0  0.000000  0.176091  0.176091   0.000000      0.0  0.000000  0.000000   
d3  0.0  0.000000  0.088046  0.088046   0.000000      0.0  0.238561  0.000000   
d4  0.0  0.477121  0.000000  0.000000   0.477121      0.0  0.000000  0.477121   
d5  0.0  0.000000  0.176091  0.176091   0.000000      0.0  0.000000  0.000000   

        gato        la     ladra    ladrar    maulla   maullar     perro   se  \
d0  0.000000  0.088046  0.000000  0.000000  0.000000  0.000000  0.150515  0.0   
d1  0.000000  0.000000  0.000000  0.778151  0.000000  0.000000  0.301030  0.0   
d2  0.000000  0.176091  0.778151  0.000000  0.000000  0.000000  0.301030  0.0   
d3  0.150515  0.088046  0.000000  0.000000  0.000000  0.000000  0.000000  0.0   
d4  0.301030  0.000000  0.000000  0.000000  0.000000  0.778151  0.000000  0.0   
d5  0.301030  0.176091  0.000000  0.000000  0.778151  0.000000  0.000000  0.0   

      y  
d0  0.0  
d1  0.0  
d2  0.0  
d3  0.0  
d4  0.0  
d5  0.0

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  
  return np.dot(v1, v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))

  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity


In [ ]:
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


#Conclusiones

![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**